controller/api/destination/watcher/profile_watcher.go

```diff
@@ -0,0 +1,234 @@
+
+	// ProfileUpdateListener is the interface that subscribers must implement.
+	ProfileUpdateListener interface {
+		Update(profile *sp.ServiceProfile)
+	}
+)
+
+// NewProfileWatcher creates a ProfileWatcher and begins watching the k8sAPI for
+// service profile changes.
+func NewProfileWatcher(k8sAPI *k8s.API, log *logging.Entry) *ProfileWatcher {
+	watcher := &ProfileWatcher{
+		profileLister: k8sAPI.SP().Lister(),
+		profiles:      make(map[ProfileID]*profilePublisher),
+		log:           log.WithField("component", "profile-watcher"),
+	}
+
+	k8sAPI.SP().Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    watcher.addProfile,
+			UpdateFunc: watcher.updateProfile,
+			DeleteFunc: watcher.deleteProfile,
+		},
+	)
+
+	return watcher
+}
+
+//////////////////////
+/// ProfileWatcher ///
+//////////////////////
+
+// Subscribe to an authority.
+// The provided listener will be updated each time the service profile for the
+// given authority is changed.
+func (pw *ProfileWatcher) Subscribe(authority string, contextToken string, listener ProfileUpdateListener) error {
+	name, err := profileID(authority, contextToken)
+	if err != nil {
+		return err
+	}
+
+	pw.log.Infof("Establishing watch on profile %s", name)
+
+	pw.profilesMu.Lock()
+	publisher, ok := pw.profiles[name]
+	if !ok {
+		profile, err := pw.profileLister.ServiceProfiles(name.Namespace).Get(name.Name)
+		if err != nil {
+			profile = nil
+		}
+
+		publisher = pw.newProfilePublisher(name, profile)
+		pw.profiles[name] = publisher
+	}
+	pw.profilesMu.Unlock()
+
+	publisher.subscribe(listener)
+	return nil
+}
+
+// Unsubscribe removes a listener from the subscribers list for this authority.
+func (pw *ProfileWatcher) Unsubscribe(authority string, contextToken string, listener ProfileUpdateListener) error {
+	name, err := profileID(authority, contextToken)
+	if err != nil {
+		return err
+	}
+	pw.log.Infof("Stopping watch on profile %s", name)
+
+	pw.profilesMu.RLock()
+	publisher, ok := pw.profiles[name]
+	pw.profilesMu.RUnlock()
+	if !ok {
+		return fmt.Errorf("cannot unsubscribe from unknown service [%s] ", name)
+	}
+	publisher.unsubscribe(listener)
+	return nil
+}
+
+func (pw *ProfileWatcher) addProfile(obj interface{}) {
+	profile := obj.(*sp.ServiceProfile)
+	id := ProfileID{
+		Namespace: profile.Namespace,
+		Name:      profile.Name,
+	}
+
+	pw.profilesMu.Lock()
+	publisher, ok := pw.profiles[id]
+	if !ok {
+		publisher = pw.newProfilePublisher(id, profile)
+		pw.profiles[id] = publisher
+
+	}
+	pw.profilesMu.Unlock()
+
+	publisher.update(profile)
+}
+
+func (pw *ProfileWatcher) updateProfile(old interface{}, new interface{}) {
+	pw.addProfile(new)
+}
+
+func (pw *ProfileWatcher) deleteProfile(obj interface{}) {
+	profile := obj.(*sp.ServiceProfile)
+	id := ProfileID{
+		Namespace: profile.Namespace,
+		Name:      profile.Name,
+	}
+
+	pw.profilesMu.RLock()
+	publisher, ok := pw.profiles[id]
+	pw.profilesMu.RUnlock()
+	if ok {
+		publisher.update(nil)
+	}
```

Comment:
Would there be an instance when `!ok` needs further action? Or can we assume that the profile was already deleted in that case?

Reply:
If the profile isn't present, we've already reached the desired state so no further action should be necessary.